File: src/main/java/engine/feature/animation/BasicAnimation.java
package engine.feature.animation;

import java.util.Map;
import java.util.TreeMap;

public class BasicAnimation {

    int currentFrameX;
    int currentFrameY;

    private static Map<String, BasicAnimation> map = new TreeMap<>();
    private String animName;

    private int animationId;
    private int usedLayerId;
    private int firstPosX;
    private int lastPosX;
    private int firstPosY;
    private int lastPosY;
    private int framesCountX;
    private int framesCountY;
    private float accumulatedTime;
    private float timeLimit;

    public BasicAnimation(final String animName, final int animationId, final int usedLayerId,
                          final int framesCountX, final int framesCountY, final float timeLimit) {
        this.animName = animName;
        this.animationId = animationId;
        this.usedLayerId = usedLayerId;

        this.framesCountX = framesCountX;
        this.framesCountY = framesCountY;

        this.currentFrameX = 1;
        this.currentFrameY = 1;

        this.firstPosX = 1;
        this.firstPosY = 1;
        this.lastPosX = framesCountX;
        this.lastPosY = framesCountY;

        this.timeLimit = timeLimit;
        this.accumulatedTime = 0f;

        addNewAnim(this);
    }

    public void changeFrame(final float deltaTime) {
        accumulatedTime += deltaTime;
        if (accumulatedTime >= timeLimit) {
            accumulatedTime = 0f;

             //System.out.println("X1: " + currentFrameX + " Y1: " + currentFrameY);
            if (framesCountX != 1) {
                if (currentFrameX + 1 > lastPosX) {
                    currentFrameX = firstPosX;

                    if (framesCountY != 1) {
                        if (currentFrameY + 1 > lastPosY)
                            currentFrameY = firstPosY;
                        else
                            currentFrameY++;
                    }
                } else
                    currentFrameX++;
            }
        }
          //System.out.println("X2: " + currentFrameX + " Y2:" + currentFrameY);
    }

    public String getName() {
        return this.animName;
    }

    public static void addNewAnim(final BasicAnimation a) {
        if (!map.containsKey(a.getName()))
            map.put(a.getName(), a);
    }

    public BasicAnimation animForName(final String animName) {
        return map.get(animName);
    }

    public void setFirstPosX(final int firstPosX) {
        this.firstPosX = firstPosX;
    }

    public void setLastPosX(final int lastPosX) {
        this.lastPosX = lastPosX;
    }

    public void setCurrentFrameX(final int currentFrameX) {
        this.currentFrameX = currentFrameX;
    }

    public void setCurrentFrameY(final int currentFrameY) {
        this.currentFrameY = currentFrameY;
    }
}

File: src/main/java/engine/feature/animation/BasicAnimation.kt
package engine.feature.animation

class BasicAnimation(val name: String,
                     private val animationId: Int,
                     private val usedLayerId: Int,
                     private val framesCountX: Int,
                     private val framesCountY: Int,
                     private val timeLimit: Float) {

    internal var currentFrameX: Int = 1
    internal var currentFrameY: Int = 1

    private var playFunction: ((Float, Int, Int, Int, Int) -> Unit)? = null

    private var lastPosX: Int = framesCountX
    private val lastPosY: Int = framesCountY

    private var firstPosX: Int = 1
    private val firstPosY: Int = 1

    private var accumulatedTime: Float = 0f

    init {
        addNewAnimToMap(this)
    }

    fun play(deltaTime: Float) {

        playFunction?.let {
            it.invoke(deltaTime, currentFrameX, currentFrameY, firstPosX, firstPosY)
            return
        }

        // Default animation sequence
        accumulatedTime += deltaTime
        if (accumulatedTime >= timeLimit) {
            accumulatedTime = 0f

            if (isMultiFramedByX()) {
                if (isLastFrameX()) {
                    currentFrameX = firstPosX

                    if (isMultiFramedByY()) {
                        if (isLastFrameY())
                            currentFrameY = firstPosY
                        else
                            currentFrameY++
                    }
                } else
                    currentFrameX++
            }
        }
    }

    private fun isLastFrameX() = currentFrameX + 1 > lastPosX

    private fun isLastFrameY() = currentFrameY + 1 > lastPosY

    private fun isMultiFramedByX() = framesCountX != 1

    private fun isMultiFramedByY() = framesCountY != 1

    fun setFirstPosX(posX: Int) {
        firstPosX = posX
    }

    fun setLastPosX(posX: Int) {
        lastPosX = posX
    }

    fun setCurrentFrameX(frame: Int) {
        currentFrameX = frame
    }

    fun setCurrentFrameY(frame: Int) {
        currentFrameY = frame
    }

    fun setPlayFunction(func: ((Float, Int, Int, Int, Int) -> Unit)?) {
        playFunction = func
    }

    companion object {
        private val map = HashMap<String, BasicAnimation>()

        fun addNewAnimToMap(a: BasicAnimation) {
            map[a.name] = a
        }

        fun animForName(animName: String): BasicAnimation? {
            return map[animName]
        }
    }
}
